torchtune/modules/model_fusion/fusion_layer.py

```diff
@@ -0,0 +1,123 @@
+    ):
+        super().__init__()
+        self.layer = layer
+        self.fusion_layer = fusion_layer
+        self.fusion_first = fusion_first
+
+        # Keep FusionLayer wrappings out of the state_dict
+        self._register_state_dict_hook(FusionLayer._state_dict_hook)
+        self._register_load_state_dict_pre_hook(
+            FusionLayer._load_state_dict_hook, with_module=True
+        )
+        # TODO: Switch to register_load_state_dict_pre_hook and
+        # register_state_dict_pre_hook after PyTorch v2.5
+
+    def _state_dict_hook(self, state_dict, *args, **kwargs):
+        """Remove "layer" from the original layer in the state_dict
+        name. This keeps the orginal state dict name for the layer
+        from before fusing with the fusion_layer.
+
+        [!Note] This update changes the order of the OrderedDict
+        """
+        keys = list(state_dict.keys())
+        for key in keys:
+            if key.startswith("layer"):
+                new_key = key.replace("layer.", "")
+                state_dict[new_key] = state_dict[key]
+                del state_dict[key]
+
+    def _load_state_dict_hook(self, state_dict, *args, **kwargs):
+        """Apply extra "layer" prefix to the state_dict key to
+        account for the FusionLayer wrapping.
+        """
+        keys = list(state_dict.keys())
+        for key in keys:
+            if not key.startswith("fusion_layer"):
+                new_key = "layer." + key
+                state_dict[new_key] = state_dict[key]
+                del state_dict[key]
+
+    def setup_cache(self, batch_size: int, dtype: torch.dtype) -> None:
+        """Setup key value cache for both layers.
+
+        Args:
+            batch_size (int): batch size for the caches.
+            dtype (torch.dtype): dtype for the caches.
+        """
+        self.layer.setup_cache(batch_size, dtype)
+        self.fusion_layer.setup_cache(batch_size, dtype)
+
+    @property
+    def cache_enabled(self) -> bool:
+        """Check if the key value caches are setup."""
+        return self.layer.cache_enabled
+
+    def reset_cache(self):
+        """Reset both layers' key value caches."""
+        self.layer.reset_cache()
+        self.fusion_layer.reset_cache()
+
+    def fusion_params(self) -> List[str]:
+        """
+        Return parameters of fusion layer.
+        """
+        fusion_params = [
+            f"fusion_layer.{k}" for k, v in self.fusion_layer.named_parameters()
+        ]
+        return fusion_params
+
+    def forward(self, x: Tensor, **kwargs: Dict) -> Tensor:
+        """
+        Args:
+            x (Tensor): input tensor with shape
+                [batch_size x seq_length x embed_dim]
+            **kwargs (Dict): all additional layer args
+
+        Returns:
+            Tensor: output tensor with same shape as input
+                [batch_size x seq_length x embed_dim]`
+
+        """
+        if self.fusion_first:
+            x = self.fusion_layer(x, **kwargs)
```

Comment:
should the kwargs be shared across both the original layer and the fusion layer? can they be different types of nn.Modules?

Reply:
This is a layer meant to be used in TransformerDecoder. So it would be expected to follow the same input/output. I'd be in favor of creating a TransformerLayer protocol though. @ebsmothers